### ISSUE TITLE
Having site editor in console. 

### ISSUE THREAD

**Issue by shirishag75** [mars-sim/mars-sim#228]
It would be nice if we could have site editor somehow in console as well. We could have just a menu with some info. that gives an idea of what the site may have or not or how difficult or dangerous the site is or not OR possibility of doing science research or not. Look forward to see how that can be done.

**Comment by mokun** [comment]
Yes. We could incorporate some key ideas of the site editor into the console. Things such as letting the player to define the latitude and longitude and naming the site are easy.

Quick question : is there any reasons you don't want a graphical interface ?

> what the site may have or not or how difficult or dangerous the site is or not OR possibility of doing science research or not.

We'll need to define what qualify or quantify the feasibility of a landing site.

I'll take that to mean that the region has too many cliffs and too much elevation change.

Currently, we rely on our 1440x2880 topographical map (~ 7 MB) to tell us at any given coordinate on Mars what the elevation is. Of course, the higher the resolution, the more accurate the data will be available, say, 3840 × 2160 (4K) would be a step better.

May be we may be open to let users to download their higher resolution/preferred height map ?

**Comment by shirishag75** [comment]
> Yes. We could incorporate some key ideas of the site editor into the console. Things such as letting the player to define the latitude and longitude and naming the site are easy.
> 
> Quick question : is there any reasons you don't want a graphical interface ?
> 
Hi Manny, 
I am not saying that at all. What I am saying is also have a site editor for those who either cannot or do not want to have a GUI. For instance, if I could get the game running in the background (passively) and after every few hrs. just check via the beryl console how things are, that would be good enough. 

As far as topographical maps are concerned, I dunno what the sizes of the map should be. For e.g. console or headless users the download should be either embedded in the game file or be fetched from outside but be small ~ 1-2 mb . As far as higher resolutions are concerned, I would suggest you to have a central place (preferably a CDN cloud cloudfront or whatever you may have) so it's accessible and available via curl/wget/https/ftp whtever you think is good. Make sure that there is a possibility of downloading diffs only if a previous version of map is there. There is git-lfs although haven't used it at all so dunno how good/bad the experience is https://git-lfs.github.com/

**Comment by shirishag75** [comment]
You could also use something like https://www.reddit.com/r/blender/comments/ce2ueg/the_proceduralize_nodes/ https://www.reddit.com/r/blender/comments/ce2ueg/the_proceduralize_nodes/. Maybe have something similar in Java, some sort of random scenario/landscape creators.

**Comment by mokun** [comment]
We may be able to use the url provided by https://www.openplanetary.org/opm/basemaps to fetch tile maps online after we have converted to JavaFX/openjfx

**Comment by Urwumpe** [comment]
What about exposing a scripting language API for Mars-sim to the console to be used for this? We could allow pausing the simulation and then manipulate it by script commands. For example, we could expose some algorithms to test the suitability of a site by various methods (topography, known minerals around it, distance to other settlements, climate).

We could maybe also allow placing new buildings and rovers that way.

**Comment by mokun** [comment]
> scripting language API

Yes. What do we need to add in order to have a scripting language API and how ?

**Comment by mokun** [comment]
@Urwumpe , the more I think about this scripting API, the more I like the idea of it.

I've thought of a few things over the years to set mars-sim on the course of becoming more scientifically useful than it is now. 

One of which is to build an interpreter language on top of it via the console. 

What if we make the console [MATLAB](https://www.youtube.com/watch?v=RnpmsjtFEcI)-like ?

> btw, I still use its open source version called [FreeMat](http://freemat.sourceforge.net/#download) often. Some use Octave but I use FreeMat mostly.

So if scientists and engineers do needs a general simulation environment to run physics and operational tests for verifying some aspects of their design in a Mars/Lunar colony, why not making mars-sim a future platform for that application ?

**Comment by mokun** [comment]
@Urwumpe  and @shirishag75 , 

I imagine that running mars-sim would generate data sets that engineers would be interested in seeing how well things would work or not.

Say I want to extract the temperature profile of a settlement between sol 1 and sol 20.

I would do the following : 

```
y = Schiaparelli Point temperature
t = sol 1 : 0.1 : 20 
plot (y, t)
```

Essentially what I would like to do is exposing the temperature of the settlement Schiaparelli Point to become a vector object called y.

I also set t to become a time vector. The *resolution* of t would become 100 millisols (since 0.1 of one sol is 100 millisols).

So, t would have the value of `[1.0, 1.1, 1.2, 1.3, 1.4, 1.5 ...19.5, 19.6, 19.7, 19.8, 19.9, 20.0]` 

Now with `plot (y,t)`, I set y to be the y-axis and t to be the x-axis and output a graph of showing a curve that illustrate how the temperature fluctuate over the course of the last 20 mission sols. 

And I can definitely see the importance of figuring out how to integrate the use of JSON for importing and exporting data as we go down this path.

**Comment by Urwumpe** [comment]
Well, since MSP is more an agent simulation and not a physical simulation, more interesting might be heatmaps of the people or how much time is spend for a task. For the rest, we are rather consuments of other simulation models than really defining something better than other simulations. For example, a weather model of Mars can be done better else where.  What this simulation really excels at is simulating multiple persons interacting. 

As basic programming language syntax, what about using Groovy there? Then we just need to export the objects that can be used for interacting with the simulation, for example a facade for creating a new scenario or a facade to extract data. 

And can't Matlab (or others) read JSON data from a webservice?

**Comment by shirishag75** [comment]
I would like to have the game simple and fun. Complex interaction and things could either be optional or add-ins. What we would need is to have tutorial mode for newbies if you look it as a game. For now, we need to know that when the spaceship lands, it has resources nearby as well as some potential to do some science which may have benefit for Martian adaption and may or may not have have benefits back home (planet 4589, earth :) ) For the site-editor in console it shouldn't even need maps but some kind of numbers which make it easy for the gamer/newbie simulator to see between two choices and make a choice based on how he wants to play. 

For e.g. some players may like high risk and high reward scenarios while others may look at long haul way of doing things. So it has to appeal to the sensitiveness of both the players. I am sure we could also look at leveling and toughness etc. to tone up things but guess that will be far into the future. For now, if we could do simple stuff, might make more sense. Let me know what you think ? 

As far as choosing groovy is concerned, it is already in Debian 2.4.17 and there is also libgmetrics-groovy-java library (0.7.4) if that is something you might be interested.

**Comment by bevans2000** [comment]
No, it would be very hard and I don't think it's worth it. Are you thinking of closing this?
Once the SIM is running it can't be changed anyway.
